src/extension.ts

```diff
@@ -120,19 +120,89 @@ export async function activate(context: vscode.ExtensionContext): Promise<vscode
+				const currentFile: string = methodLevelTarget.substring(0, methodLevelTarget.lastIndexOf('#'));
+				const projectDir: string = targeting.getProjectDir(currentFile);
+
 				return _runAndDisplayDfa(context, {
 					commandName: Constants.COMMAND_RUN_DFA_ON_SELECTED_METHOD,
 					outputChannel
-				}, customCancellationToken);
+				}, customCancellationToken, methodLevelTarget, projectDir);
 			});
 		}
 	});
-	context.subscriptions.push(runOnActiveFile, runOnSelected, runDfaOnSelectedMethod, removeDiagnosticsOnActiveFile, removeDiagnosticsOnSelectedFile, removeDiagnosticsInRange);
+
+	const runDfaOnWorkspaceCmd = vscode.commands.registerCommand(Constants.COMMAND_RUN_DFA, async () => {
+		await _runDfa(context, outputChannel);
+	});
+	context.subscriptions.push(runOnActiveFile, runOnSelected, runDfaOnSelectedMethodCmd, runDfaOnWorkspaceCmd, removeDiagnosticsOnActiveFile, removeDiagnosticsOnSelectedFile, removeDiagnosticsInRange);
 	TelemetryService.sendExtensionActivationEvent(extensionHrStart);
 	outputChannel.appendLine(`Extension sfdx-code-analyzer-vscode activated.`);
 	return Promise.resolve(context);
 }
 
+async function _runDfa(context: vscode.ExtensionContext, outputChannel: vscode.LogOutputChannel) {
+	const choice = await vscode.window.showQuickPick(
+		["***Yes***", "***No***"],
+		{
+			placeHolder: "***Re-run previously failed violations only?***",
+			canPickMany: false,
+			ignoreFocusOut: true
+		}
+	);
+
+	// Default to "Yes" if no choice is made
+	const rerunFailedOnly = choice !== "Yes";
+
+	if (!rerunFailedOnly) {
+		void vscode.window.showWarningMessage('***A full run of the graph engine will happen in the background. You can cancel this by clicking on the status progress.***');
+		await runDfaOnWorkspace(context, outputChannel);
+	} else if (!violationsCacheExists()) {
+		void vscode.window.showWarningMessage('***A full run of the graph engine will happen in the background since no existing cache found. You can cancel this by clicking on the status progress.***');
+		await runDfaOnWorkspace(context, outputChannel);
+	} else {
+		// Do nothing for now. This will be implemented as part of W-15639759
+	}
+}
+
+async function runDfaOnWorkspace(context: vscode.ExtensionContext, outputChannel: vscode.LogOutputChannel) {
+	await vscode.window.withProgress({
+		location: vscode.ProgressLocation.Window,
+		title: messages.graphEngine.spinnerText,
+		cancellable: true
+	}, async (progress, token) => {
+		token.onCancellationRequested(async () => {
+			await _stopExistingDfaRun(context, outputChannel);
+		});
+		customCancellationToken = new vscode.CancellationTokenSource();
+		customCancellationToken.token.onCancellationRequested(async () => {
+			customCancellationToken?.dispose();
+			customCancellationToken = null;
+			await vscode.window.showInformationMessage(messages.graphEngine.noViolationsFound);
+			return;
+		});
+
+		const projectDir: string[] = vscode.workspace.workspaceFolders?.map(folder => folder.uri.path);
```

Comment:
workspaceRoot is now deprecated and this seems to be the way to get the project directory's root. I couldn't get much info on this from the docs, but I have tested this and it works as expected.

Reply:
Should we move this logic inside of targeting.getProjectDir() and make currentFile an optional input argument?

Reply:
That's good idea. Done.